packages/eslint-plugin/src/rules/enum-utils/shared.ts

```diff
@@ -33,8 +50,40 @@ export function getEnumTypes(
   typeChecker: ts.TypeChecker,
   type: ts.Type,
 ): ts.Type[] {
-  return tsutils
-    .unionTypeParts(type)
-    .filter(subType => isTypeFlagSet(subType, ts.TypeFlags.EnumLiteral))
-    .map(type => getBaseEnumType(typeChecker, type));
+  return getEnumLiterals(type).map(type => getBaseEnumType(typeChecker, type));
+}
+
+/**
+ * Returns the enum key that matches the given literal node, or null if none
+ * match. For example:
+ * ```ts
+ * enum Fruit {
+ *   Apple = 'apple',
+ *   Banana = 'banana',
+ * }
+ *
+ * getEnumKeyForLiteral([Fruit.Apple, Fruit.Banana], 'apple') --> 'Fruit.Apple'
+ * getEnumKeyForLiteral([Fruit.Apple, Fruit.Banana], 'banana') --> 'Fruit.Banana'
+ * getEnumKeyForLiteral([Fruit.Apple, Fruit.Banana], 'cherry') --> null
+ * ```
+ */
+export function getEnumKeyForLiteral(
+  enumLiterals: ts.LiteralType[],
+  literal: unknown,
+): string | null {
+  for (const enumLiteral of enumLiterals) {
+    if (enumLiteral.value === literal) {
+      const { symbol } = enumLiteral;
+
+      const memberDeclaration = symbol.valueDeclaration as ts.EnumMember;
+      const enumDeclaration = memberDeclaration.parent;
+
+      const enumName = enumDeclaration.name.getText();
+      const memberName = memberDeclaration.name.getText();
```

Comment:
[Testing] Although TypeScript says its enums can't have computed property names, `["key"]` literals are allowed. There should be some tests for that. I see you live added one for `'a' = 1,`, let's do `['a'] = 1,` (or an equivalent) too? And `['a' /* gotcha */] = 1,` (or an equivalent) - comments tend to get in the way when using `getText()`!

Fortunately, `[someVariableId]` and `[123]` aren't allowed. So if you need an assertion or two, that might make sense.

Reply:
Done